src/core/AutomationPattern.cpp

```diff
@@ -460,69 +653,86 @@ void AutomationPattern::flipY()
 
 
 
-void AutomationPattern::flipX( int length )
+void AutomationPattern::flipX(int length)
 {
+	QMutexLocker m(&m_patternMutex);
+
+	timeMap::const_iterator it = m_timeMap.lowerBound(0);
+
+	if (it == m_timeMap.end()) { return; }
+
+	// Temporary map where we will store the flipped version
+	// of our pattern
 	timeMap tempMap;
 
-	timeMap::ConstIterator iterate = m_timeMap.lowerBound(0);
 	float tempValue = 0;
-	int numPoints = 0;
+	float tempOutValue = 0;
 
-	for( int i = 0; ( iterate + i + 1 ) != m_timeMap.end() && ( iterate + i ) != m_timeMap.end() ; i++)
-	{
-		numPoints++;
-	}
-
-	float realLength = ( iterate + numPoints ).key();
+	// We know the QMap isn't empty, making this safe:
+	float realLength = m_timeMap.lastKey();
 
-	if ( length != -1 && length != realLength)
+	// If we have a positive length, we want to flip the area covered by that
+	// length, even if it goes beyond the pattern. A negative length means that
+	// we just want to flip the nodes we have
+	if (length >= 0 && length != realLength)
 	{
-		if ( realLength < length )
+		// If length to be flipped is bigger than the real length
+		if (realLength < length)
 		{
-			tempValue = valueAt( ( iterate + numPoints ).key() );
-			putValue( TimePos( length ) , tempValue, false);
-			numPoints++;
-			for( int i = 0; i <= numPoints; i++ )
+			// We are flipping an area that goes beyond the last node. So we add a node to the
+			// beginning of the flipped timeMap representing the value of the end of the area
+			tempValue = valueAt(length);
+			tempMap[0] = AutomationNode(this, tempValue, 0);
+
+			// Now flip the nodes we have in relation to the length
+			do
 			{
-				tempValue = valueAt( ( iterate + i ).key() );
-				TimePos newTime = TimePos( length - ( iterate + i ).key() );
-				tempMap[newTime] = tempValue;
-			}
+				tempValue = INVAL(it);
+				tempOutValue = OUTVAL(it);
+				TimePos newTime = TimePos(length - POS(it));
+
+				tempMap[newTime] = AutomationNode(this, tempValue, tempOutValue, newTime);
+
```

Comment:
You should swap in and out values to really flip the pattern. The same for 2 more cases below.